internal/util/update/fastforward.go

```diff
@@ -43,45 +50,197 @@ func (u FastForwardUpdater) Update(options UpdateOptions) error {
 	g := options.KptFile.Upstream.Git
 	g.Ref = options.ToRef
 	g.Repo = options.ToRepo
-	if err := errorIfChanged(g, options.PackagePath); err != nil {
+
+	defaultRef, err := gitutil.DefaultRef(g.Repo)
+	if err != nil {
 		return err
 	}
 
-	// refetch the package
-	return get.Command{Destination: options.PackagePath, Clean: true, Git: g}.Run()
-}
+	// get the original repo
+	original := &git.RepoSpec{OrgRepo: g.Repo, Path: g.Directory, Ref: g.Commit}
+	if err := get.ClonerUsingGitExec(original, defaultRef); err != nil {
+		return errors.Errorf("failed to clone git repo: original source: %v", err)
+	}
+	defer os.RemoveAll(original.AbsPath())
 
-// errorIfChanged returns an error if the package at pkgPath has changed from the upstream
-// source referenced by g.
-func errorIfChanged(g kptfile.Git, pkgPath string) error {
-	original := &git.RepoSpec{
-		OrgRepo: g.Repo,
-		Path:    g.Directory,
-		Ref:     g.Commit,
+	// get the updated repo
+	updated := &git.RepoSpec{OrgRepo: options.ToRepo, Path: g.Directory, Ref: options.ToRef}
+	if err := get.ClonerUsingGitExec(updated, defaultRef); err != nil {
+		return errors.Errorf("failed to clone git repo: updated source: %v", err)
 	}
-	defaultRef, err := gitutil.DefaultRef(g.Repo)
+	defer os.RemoveAll(updated.AbsPath())
+
+	commit, err := lookupCommit(updated.AbsPath())
 	if err != nil {
 		return err
 	}
-	err = get.ClonerUsingGitExec(original, defaultRef)
+
+	// Verify that there are no local changes that would prevent us from
+	// using the FastForward strategy.
+	if err := u.checkForLocalChanges(options.AbsPackagePath, original.AbsPath(), options.KptFile); err != nil {
+		return err
+	}
+
+	// Look up all subpackages across the local package and the updated (from upstream)
+	// package.
+	subPkgPaths, err := findAllSubpackages(options.AbsPackagePath, updated.AbsPath())
 	if err != nil {
-		return errors.Errorf("failed cloning git repo: %v", err)
+		return err
 	}
-	defer os.RemoveAll(original.Dir)
-	diff, err := copyutil.Diff(original.AbsPath(), pkgPath)
+
+	// Update each package individually, starting with the root package.
```

Comment:
Most of the logic here seems to be common for all update strategies. Should we push it into library and reuse the logic. Or make it a separate method at the least.